marshmallow/fields.py

```diff
@@ -1195,31 +1195,23 @@ def _deserialize(self, value, attr, data):
         if not self.value_container and not self.key_container:
             return value
 
-        errors = {}
+        errors = collections.defaultdict(dict)
         values = list(value.values())
         keys = list(value.keys())
         if self.key_container:
             for idx, key in enumerate(keys):
                 try:
                     keys[idx] = self.key_container.deserialize(key)
                 except ValidationError as e:
-                    errors[key] = [
-                        'Invalid key: {}'.format(message)
-                        for message in e.messages
-                    ]
+                    errors[key]['key'] = e.messages
         if self.value_container:
             for idx, item in enumerate(values):
                 try:
                     values[idx] = self.value_container.deserialize(item)
                 except ValidationError as e:
                     values[idx] = e.data
                     key = keys[idx]
-                    if key not in errors:
-                        errors[key] = []
-                    errors[key].extend([
-                        'Invalid value: {}'.format(message)
-                        for message in e.messages
```

Comment:
Er, wait. The `messages` property of a `ValidationError` can be a "string, list, or dictionary of error messages".

https://marshmallow.readthedocs.io/en/latest/api_reference.html#marshmallow.exceptions.ValidationError.messages

Ya, I was assuming it was a list. We can't safely merge the value and key errors, so making it a dict to separate them makes sense.